build.assets/tooling/cmd/tag-release/main.go

```diff
@@ -0,0 +1,227 @@
+			Message: annotation,
+		}
+	}
+
+	for _, repo := range repos {
+		w, err := repo.Worktree()
+		if err != nil {
+			if !force {
+				return trace.Wrap(err, "Failed extracting worktree")
+			}
+			log.Printf("WARNING: failed extracting woirktree")
+			continue
+		}
+
+		log.Printf("Tagging %s", w.Filesystem.Root())
+
+		tagExists, err := statTag(repo, tag)
+		if err != nil {
+			return trace.Wrap(err, "failed enumerating tags")
+		}
+
+		if tagExists {
+			if !force {
+				return trace.BadParameter("tag already exists")
+			}
+			log.Printf("WARNING: tag \"%s\" already exists. Deleting.", tag)
+			err = repo.DeleteTag(tag)
+			if err != nil {
+				log.Printf("WARNING: Faled to delete tag: %s", err)
+				continue
+			}
+		}
+
+		head, err := repo.Head()
+		if err != nil {
+			return trace.Wrap(err, "failed fetching head revision")
+		}
+
+		log.Printf("Tagging %s => %s", w.Filesystem.Root(), head.String())
+		_, err = repo.CreateTag(tag, head.Hash(), opts)
+		if err != nil {
+			if !force {
+				return trace.Wrap(err, "")
+			}
+			log.Printf("Failed to set tag: %s", err)
+		}
+	}
+
+	return nil
+}
+
+// pushTags pushed the created tags to all of the listed submodules.
+//
+// The default behaviour is to halt at the first error, but if force is `true`,
+// then `pushTags` will ignore some errors and try to continue. Ignored errors
+// include:
+//   - Failure to extract repo metadata
+//   - failure to push to a repo.
+//
+// When an operation on a reposiotory failes (and force is `true`), `pushTags`
+// will give up on the current repsitory and move on to the next one.
+func pushTags(repos []*git.Repository, remote string, tag string, force bool) error {
+	refname := path.Join("refs", "tags", tag)
+	refspec := fmt.Sprintf("%s:%s", refname, refname)
+
+	for _, repo := range repos {
+		w, err := repo.Worktree()
+		if err != nil {
+			if !force {
+				return trace.Wrap(err, "Failed extracting worktree")
+			}
+			log.Printf("WARNING: failed extracting woirktree")
+			continue
+		}
+
+		log.Printf("Pushing: %s", w.Filesystem.Root())
+
+		opts := git.PushOptions{
+			RemoteName: remote,
+			RefSpecs:   []config.RefSpec{config.RefSpec(refspec)},
+		}
+
+		if err := repo.Push(&opts); err != nil {
+			if !force {
+				return trace.Wrap(err, "failed pushing tags")
+			}
+			log.Printf("WARNING: failed pushing tag: %s", err)
+		}
+	}
+
+	return nil
+}
+
+// statTag checks to see if a tag with the given name already exists in a
+// repository, either as a simple lightweight tag, or a full annotated tag.
+func statTag(repo *git.Repository, tag string) (bool, error) {
+	tagpath := "refs/tags/" + tag
+	res := false
+	simpleTags, err := repo.Tags()
+	if err != nil {
+		return false, trace.Wrap(err, "failed listing tags")
+	}
+	simpleTags.ForEach(func(t *plumbing.Reference) error {
+		if t.Name().String() == tagpath {
+			res = true
+			return nil
+		}
+		return nil
+	})
+
+	if err != nil {
+		return false, trace.Wrap(err, "Failed enumerating tags")
+	}
```

Comment:
Need to assign err above
```suggestion
	err = simpleTags.ForEach(func(t *plumbing.Reference) error {
		if t.Name().String() == tagpath {
			res = true
			return nil
		}
		return nil
	})

	if err != nil {
		return false, trace.Wrap(err, "Failed enumerating tags")
	}
```

Reply:
although by the looks of it, error will always be `nil`? Might be able to just delete the error check if that's the case.

Reply:
I expect that the underlying enumeration can fail, even if our callbacks don't